### PR TITLE
Fixed Ship and Cargo classes 

### LINKED ISSUE #30: Fixed overloading operator += and -= in (Cargo and Ship classes)
What happened when we reach the limit of capacity or drop below 0? You need to handle error cases. Maybe throw exception, maybe return std::nullopt or nullptr maybe set some specific value like std::numeric_limits::min/max.

### LINKED ISSUE #31: Applied const ref in places where std::string is passed as arg or is return value
we should check getters, setters, constructors...

### PR DESCRIPTION
**Pull request by Martyna-Tu**
Ship:
changed order, added const ref to std::string in methods and c-tor, added alias CargoVec, forward declaration
Cargo: 
changed order, added const ref to std::string in methods and c-tor

### CODE REVIEW

File: shm/Ship.hpp [hunk 1]
```diff
@@ -2,31 +2,34 @@
...
 class Ship {
+public:
+    using CargoVec = std::vector<std::unique_ptr<Cargo>>;
+    Ship();
+    Ship(size_t capacity, size_t maxCrew, size_t speed, const std::string& name, const size_t id, CargoVec cargoVec);
+    Ship(size_t maxCrew, size_t speed, size_t id);
+ 
+    void setName(const std::string& name);
+ 
+    Ship& operator-=(size_t num);
```
Comment (runer53248) [inline shm/Ship.hpp]: Don't check if go below 0. 
Parameter num don't give meaning of what we add or substract.
Maybe should change operator for meaningful  addCrew().

File: shm/Cargo.hpp [hunk 2]
```diff
@@ -3,20 +3,19 @@
...
 public:
-    Cargo(std::string name, size_t amount, size_t basePrice);
+    Cargo(const std::string& name, size_t amount, size_t basePrice);
 
     Cargo& operator+=(size_t amount);
     Cargo& operator-=(size_t amount);
     bool operator==(const Cargo& other) const;
 
-    std::string getName() const;
+    const std::string& getName() const;
```
Comment (runer53248) [inline shm/Cargo.hpp:14]: If cargo will be destroyed we can have dangling reference to its name.

File: shm/Ship.cpp [hunk 3]
```diff
@@ -1,20 +1,15 @@
+#include "Cargo.hpp"
 #include "Ship.hpp"
-    
-Ship::Ship() 
-    : id_(-1)
+
+Ship::Ship()
+    : Ship(0, 0, 0, "", -1, {})
```
Comment (runer53248) [inline shm/Ship.cpp]: Magic numbers.
Better if initialize these parameters in declarations.

**Review by gyrydymptum — APPROVED** [review]

File: shm/Ship.cpp [hunk 4]
```diff
@@ -1,20 +1,15 @@
...
-    , speed_(speed)
-    , name_(name)
-    , id_(id)
-    , cargoVec_(std::move(cargoVec))
+Ship::Ship(size_t capacity, size_t maxCrew, size_t speed, const std::string& name, const size_t id, CargoVec cargoVec)
+    : capacity_(capacity), maxCrew_(maxCrew), crew_(0), speed_(speed), name_(name), id_(id), cargoVec_(std::move(cargoVec))
 {}
 
-Ship::Ship(int maxCrew, int speed, size_t id)
+Ship::Ship(size_t maxCrew, size_t speed,const size_t id)
```
Comment (runer53248) [inline shm/Ship.cpp:8]: Unnecesary const in id argument.
If id don't define some ship type and shouldn't be duplicated consider static id counter in class for secure different values in them.

File: shm/Ship.hpp [hunk 5]
```diff
@@ -2,31 +2,34 @@
 #include <memory>
 #include <string>
 #include <vector>
-#include "Cargo.hpp"
+
+class Cargo;
```
Comment (runer53248) [inline shm/Ship.hpp:6]: Nice.

File: shm/Cargo.cpp [hunk 6]
```diff
@@ -10,15 +12,18 @@ Cargo& Cargo::operator+=(size_t amount) {
 }
 
 Cargo& Cargo::operator-=(size_t amount) {
+    if (amount_ < amount) {
+        throw std::out_of_range("not allow! you will be below zero!");
```
Comment (Rogal255) [inline shm/Cargo.cpp]: English grammar fix - should be "Not allowed! (...)". Also consider adding more descriptive error message, explicitly mention that CARGO can't have negative amount. Maybe adding which method has thrown an exception would be helpful during future debugging.

File: shm/Ship.cpp [hunk 7]
```diff
@@ -1,35 +1,31 @@
...
-
 void Ship::setName(const std::string& name) { 
     name_ = name; 
 }
 
-Ship& Ship::operator-=(size_t num) {
-    crew_ -= num;
+Ship& Ship::operator-=(size_t numCrew) {
+    if (crew_ < numCrew) {
+        throw std::out_of_range("not allow! you will be below zero!");
```
Comment (Rogal255) [inline shm/Ship.cpp]: English grammar fix - should be "Not allowed! (...)". Also consider adding more descriptive error message, explicitly mention that ship crew can't have negative value. Maybe adding which method has thrown an exception would be helpful during future debugging.

File: shm/Ship.cpp [hunk 8]
```diff
@@ -1,35 +1,31 @@
...
+    }
+    crew_ -= numCrew;
     return *this;
 }
 
-Ship& Ship::operator+=(size_t num) {
-    crew_ += num;
+Ship& Ship::operator+=(size_t numCrew) {
+    if (crew_ + numCrew > maxCrew_) {
+        throw std::out_of_range("not allow! there will be too many people on the board!");
```
Comment (Rogal255) [inline shm/Ship.cpp]: English grammar fix - should be "Not allowed! (...)". Maybe adding which method has thrown an exception would be helpful during future debugging.

**Review by Rogal255 — CHANGES_REQUESTED** [review]

File: shm/Ship.cpp (hunk 4 above)
Comment (Rogal255) [reply shm/Ship.cpp:8]: Good idea with a static counter. In that case you will need to handle id initialization in constructor. Something like id_ {++s_counter} in init list. Where s_counter is static variable representing last assigned id_.

File: shm/Ship.cpp (hunk 7 above)
Comment (runer53248) [inline shm/Ship.cpp]: Should be std::invalid_argument.

File: shm/Ship.cpp [hunk 9]
```diff
@@ -1,35 +1,32 @@
...
+    }
+    crew_ -= numCrew;
     return *this;
 }
-
-Ship& Ship::operator+=(size_t num) {
-    crew_ += num;
+//This operator can throw exception - we have to remember that method which will use this operator, should catch it!
+Ship& Ship::operator+=(size_t numCrew) {
+    if (crew_ + numCrew > maxCrew_) {
```
Comment (runer53248) [inline shm/Ship.cpp]: Need check for unsigned value wrapping.

**Review by Rogal255 — APPROVED** [review]

**Review by runer53248 — APPROVED** [review]
Just remove a.out and be fine.

**Pull request merged**